frex/utils/constraint_solver.py

```diff
@@ -0,0 +1,208 @@
+                )
+                self.solver.Add(psc.constraint_type(ss, psc.constraint_val))
+
+        # constraints to apply to the overall solution, based on certain field names
+        # constraints are added to the solver here for similar reasons as the section_constraints
+        for oc in self.overall_constraints:
+            attributes_of_interest.add(oc.attribute_name)
+            ss = self.solver.Sum(
+                [
+                    rgetattr(self.candidates[i].domain_object, oc.attribute_name) * solve_choice[i, j]
+                    for i in range(candidate_count)
+                    for j in range(self.sections)
+                ]
+            )
+            self.solver.Add(oc.constraint_type(ss, oc.constraint_val))
+
+        # maximize score
+        objective_terms = []
+        for i in range(candidate_count):
+            for j in range(self.sections):
+                # in the future, maybe incorporate more into this objective function
+                # e.g., we would prefer the combination of all items in a section to have some field value in a range,
+                # so incorporate that into the score somehow?
+                objective_terms.append(self.candidates[i].total_score * solve_choice[i, j])
+        self.solver.Maximize(self.solver.Sum(objective_terms))
+
+        status = self.solver.Solve()
+
+        if status == pywraplp.Solver.OPTIMAL or status == pywraplp.Solver.FEASIBLE:
+            sections = []
+            overall_attributes = {attr:0 for attr in attributes_of_interest}
+            for j in range(self.sections):
+                section_candidates = []
+                section_attributes = {attr:0 for attr in attributes_of_interest}
+                section_score = 0
+                for i in range(candidate_count):
+                    if solve_choice[i, j].solution_value() > 0.5:
+                        # print("Candidate ", i, " assigned to section ", j)
+                        section_candidates.append(self.candidates[i])
+                        section_score += self.candidates[i].total_score
+                        for attr in attributes_of_interest:
+                            section_attributes[attr] += rgetattr(self.candidates[i].domain_object, attr)
+                sections.append(ConstraintSectionSolution(
+                    section_candidates=tuple(section_candidates),
+                    section_score=section_score,
+                    section_attribute_values=section_attributes
+                ))
+                for attr in attributes_of_interest:
+                    overall_attributes[attr] += section_attributes[attr]
+
+            return ConstraintSolution(
+                sections=tuple(sections),
+                overall_score=self.solver.Objective().Value(),
+                overall_attribute_values=overall_attributes
+            )
+        else:
+            # print('No solution found.')
+            return ()  # TODO: what to return if no solution?
```

Comment:
If it's an exceptional condition, raise an exception. Otherwise make the return type Optional[ConstraintSolution] and return None.

Either way, don't use else here. Just have 
```
if status != pywraplp.Solver.OPTIMAL and status != pywraplp.Solver.FEASIBLE:`
    # Exit in the appropriate way
```

Then the current if branch is dedented.